nifi-nar-bundles/nifi-redis-bundle/nifi-redis-extensions/src/main/java/org/apache/nifi/redis/state/RedisStateProvider.java

```diff
@@ -0,0 +1,271 @@
+            redisConnection.watch(key);
+
+            final long prevVersion = oldValue == null ? -1L : oldValue.getVersion();
+
+            final byte[] currValue = redisConnection.get(key);
+            final RedisStateMap currStateMap = serDe.deserialize(currValue);
+            final long currVersion = currStateMap == null ? -1L : currStateMap.getVersion();
+
+            // the replace API expects that you can't call replace on a non-existing value, so unwatch and return
+            if (!allowReplaceMissing && currVersion == -1) {
+                redisConnection.unwatch();
+                return false;
+            }
+
+            // start a transaction
+            redisConnection.multi();
+
+            // compare-and-set
+            if (prevVersion == currVersion) {
+                // build the new RedisStateMap incrementing the version, using latest encoding, and using the passed in values
+                final RedisStateMap newStateMap = new RedisStateMap.Builder()
+                        .version(currVersion + 1)
+                        .encodingVersion(ENCODING_VERSION)
+                        .stateValues(newValue)
+                        .build();
+
+                // if we use set(k, newVal) then the results list will always have size == 0 b/c when convertPipelineAndTxResults is set to true,
+                // status responses like "OK" are skipped over, so by using getSet we can rely on the results list to know if the transaction succeeded
+                redisConnection.getSet(key, serDe.serialize(newStateMap));
+            }
+
+            // execute the transaction
+            final List<Object> results = redisConnection.exec();
+
+            // if we have a result then the replace succeeded
+            if (results.size() > 0) {
+                replaced = true;
+            }
+
+            return replaced;
+        });
+    }
+
+    @Override
+    public void clear(final String componentId) throws IOException {
+        final StateMap currStateMap = getState(componentId);
+
+        int attempted = 0;
+        boolean updated = false;
+
+        while (!updated && attempted < 20) {
+            updated = replace(currStateMap, Collections.emptyMap(), componentId, true);
+            attempted++;
+        }
+
+        if (!updated) {
+            throw new IOException("Unable to update state due to concurrent modifications");
+        }
+    }
+
+    @Override
+    public void onComponentRemoved(final String componentId) throws IOException {
+        withConnection(redisConnection -> {
+            final byte[] key = getComponentPath(componentId).getBytes(StandardCharsets.UTF_8);
+            redisConnection.del(key);
+            return true;
+        });
+    }
+
+    @Override
+    public Scope[] getSupportedScopes() {
+        return new Scope[] {Scope.CLUSTER};
+    }
+
+    private String getComponentPath(final String componentId) {
```

Comment:
I like that idea, I added a property called "Key Prefix" since with Redis it isn't as much of a path as it is with ZooKeeper. Even without the Key Prefix they could still use the Database Index to segregate the keys, but I still like making the prefix configurable.